### DISCUSSION TITLE
Supabase's Own Schemas

### DISCUSSION THREAD

**Discussion by accmltr** [supabase/supabase#23263] [Questions] [answered]
Hi, where can I find the definitions for the built-in schemas for any new Supabase project? I'm trying to write a client library in Scala and would like a solid reference for how sessions, users etc. are defined.

**Comment by accmltr** [comment]
I often see references like this in the documentation:

![image](https://github.com/supabase/supabase/assets/19354678/ffd273ca-8ab7-42de-8c8c-f188e889713c)

But I cannot find `auth.sessions` table anywhere in my project or elsewhere online, and the same goes for any other such mentions in the docs.

**Reply by GaryAustin1** [reply, marked as answer]
You can go to the auth repository for instance and see the SQL code that it sets up.
There is no definition of the internal tables that I have seen in form of a published doc.

Normally you would look to the REST API interface for clients and not use the schemas themselves.
This has the API calls for auth towards the bottom half of the readme.  https://github.com/supabase/auth
You could also pick an existing client and look at the source for the calls it makes.

And auth.sessions is in the auth schema and can be seen using the table UI.

**Reply by accmltr** [reply]
I have been working off of the other client code, but wanted to learn how to find the right stuff myself and also be sure I'm not missing some features that might not be included in some of the client libraries.

Thank you very much for the help, appreciate it.

**Reply by GaryAustin1** [reply]
Just remember the schemas tables and columns change regularly without a major release to the REST APIs themselves.  So unless used in a public document (and I've seen mentioned a primary key) things can change (and have changed) out from under you.

**Reply by accmltr** [reply]
You're right, the models and tables defined seem to be different from the actual responses provided by the REST API.

I am still a bit confused however, since the returns shown in the Endpoints section of the readme also don't match up exactly to what I get from using them.

E.g. 

```
basicRequest
      .post(uri"$baseUri/auth/v1/token?grant_type=password")
      .headers(headers)
      .body(credentials)
      .send(backend)
```

returns:

```
{
  "access_token":"...",
  "token_type":"bearer",
  "expires_in":3600,
  "expires_at":1714101078,
  "refresh_token":"m6ybzZi6W8XaK5Hwt0ASkw",
  "user":{...}
}
```

but the readme says posting to `/token?grant_type=password` returns:

```
{
  "access_token": "jwt-token-representing-the-user",
  "token_type": "bearer",
  "expires_in": 3600,
  "refresh_token": "a-refresh-token"
}
```

I feel like I'm in a bit over my head, since I'm quite new to RESP APIs. I also couldn't figure out exactly how the responses defined in `auth/internal/api` work, but they do start making sense the more I read through them.

E.g. `auth/internal/api/token.go` snippet:

```
func (r *AccessTokenResponse) AsRedirectURL(redirectURL string, extraParams url.Values) string {
	extraParams.Set("access_token", r.Token)
	extraParams.Set("token_type", r.TokenType)
	extraParams.Set("expires_in", strconv.Itoa(r.ExpiresIn))
	extraParams.Set("expires_at", strconv.FormatInt(r.ExpiresAt, 10))
	extraParams.Set("refresh_token", r.RefreshToken)

	return redirectURL + "#" + extraParams.Encode()
}
```
Helped me understand the response a bit more.

In the end I might just use the other community clients for reference instead, as you recommended earlier.

**Reply by GaryAustin1** [reply]
I'll be honest, I have little confidence that the readme always reflects the exact state of the API, but that is my outside opinion based on dealing with the docs in general helping users and how fast they (Supabase) get things updated after changes on existing examples.  I am not commenting directly on that readme, other than that is what they provide.  I've seen it gives the basic info when helping debugging user issues. 

In general people making REST API calls directly without the clients can get things done, but IMO looking at the supabase-js client and its parts (auth-js, etc) is critical if you are trying to roll your own, as that must work with any new change.  Or at least work by the time the feaure is launched.   There are times the REST API's are weeks ahead of the actual clients using those features.   The most well documented REST API is the database from PostgREST.org but the postrest-js client could be weeks behind that based on when Supabase decides to roll in the new features.

**Reply by accmltr** [reply]
I'll just use the supabase-js client then. Thank you again for the help, very much appreciated.

**Comment by accmltr** [comment]
Just found the "Schema Visualizer" at `https://supabase.com/dashboard/project/[YOUR PROJECT ID]/database/schemas`. It lets you see the built-in Supabase schemas!

![image](https://github.com/supabase/supabase/assets/19354678/93c69158-72e8-47f3-9b4e-d404e6f5a6d2)

Hope this helps anyone else who might be looking for a place to see what type of data you might be working with.
